File: server/redis-client.js
const { createClient, commandOptions } = require("redis");

// Redis client configuration
// Support both REDIS_URL and UPSTASH_REDIS_URL (Heroku addon)
const REDIS_URL = process.env.REDIS_URL || "redis://localhost:6379";
const REDIS_TTL = 60 * 60; // 1 hour in seconds

let client = null;
let isConnected = false;

/**
 * Initialize Redis client
 */
async function initRedis() {
  if (client) {
    return client;
  }

  try {
    client = createClient({
      url: REDIS_URL,
      socket: {
        rejectUnauthorized: false, // Add this line
        reconnectStrategy: (retries) => {
          if (retries > 10) {
            console.error("Redis: Max reconnection attempts reached");
            return new Error("Max reconnection attempts reached");
          }
          const delay = Math.min(retries * 100, 3000);
          console.log(`Redis: Reconnecting in ${delay}ms...`);
          return delay;
        },
      },
    });

    client.on("error", (err) => {
      console.error("Redis Client Error:", err);
      isConnected = false;
    });

    client.on("connect", () => {
      console.log("Redis: Connecting...");
    });

    client.on("ready", () => {
      console.log("Redis: Connected and ready");
      isConnected = true;
    });

    client.on("reconnecting", () => {
      console.log("Redis: Reconnecting...");
      isConnected = false;
    });

    client.on("end", () => {
      console.log("Redis: Connection closed");
      isConnected = false;
    });

    await client.connect();
    return client;
  } catch (error) {
    console.error("Failed to initialize Redis:", error);
    throw error;
  }
}

/**
 * Get Redis client (ensures initialized)
 */
async function getClient() {
  if (!client || !isConnected) {
    await initRedis();
  }
  return client;
}

/**
 * Store wind data in Redis (with automatic chunking for large datasets)
 * @param {Object|Array} data - Wind data to store
 * @param {string} key - Redis key (default: 'wind:data')
 */
async function setWindData(data, key = "wind:data") {
  try {
    const redis = await getClient();
    const dataString = JSON.stringify(data);
    const dataSize = Buffer.byteLength(dataString, "utf8");

    // Upstash Redis free tier limit: 10 MB per request
    const MAX_SIZE = 8 * 1024 * 1024; // 8 MB to be safe

    if (dataSize > MAX_SIZE) {
      // Check if data is an array or object with large array property
      if (Array.isArray(data)) {
        // Split large array into chunks
        console.log(
          `Redis: Array too large (${dataSize} bytes), splitting into chunks...`,
        );

        const chunkSize = Math.ceil(
          data.length / Math.ceil(dataSize / MAX_SIZE),
        );
        const chunks = [];

        for (let i = 0; i < data.length; i += chunkSize) {
          chunks.push(data.slice(i, i + chunkSize));
        }

        // Store chunk count
        await redis.setEx(`${key}:chunks`, REDIS_TTL, chunks.length.toString());

        // Store each chunk
        for (let i = 0; i < chunks.length; i++) {
          await redis.setEx(
            `${key}:chunk:${i}`,
            REDIS_TTL,
            JSON.stringify(chunks[i]),
          );
        }

        console.log(
          `Redis: Stored ${data.length} items in ${chunks.length} chunks at key '${key}' with TTL ${REDIS_TTL}s`,
        );
      } else if (
        typeof data === "object" &&
        data !== null &&
        data.points &&
        Array.isArray(data.points)
      ) {
        // Object with large 'points' array - split points into chunks
        console.log(
          `Redis: Object with large points array (${dataSize} bytes), splitting points into chunks...`,
        );

        const points = data.points;
        const otherData = { ...data };
        delete otherData.points;

        const chunkSize = Math.ceil(
          points.length / Math.ceil(dataSize / MAX_SIZE),
        );
        const chunks = [];

        for (let i = 0; i < points.length; i += chunkSize) {
          chunks.push(points.slice(i, i + chunkSize));
        }

        // Store metadata (without points)
        await redis.setEx(`${key}:meta`, REDIS_TTL, JSON.stringify(otherData));

        // Store chunk count
        await redis.setEx(`${key}:chunks`, REDIS_TTL, chunks.length.toString());

        // Store each chunk
        for (let i = 0; i < chunks.length; i++) {
          await redis.setEx(
            `${key}:chunk:${i}`,
            REDIS_TTL,
            JSON.stringify(chunks[i]),
          );
        }

        console.log(
          `Redis: Stored ${points.length} points in ${chunks.length} chunks at key '${key}' with TTL ${REDIS_TTL}s`,
        );
      } else {
        throw new Error(
          `Data too large (${dataSize} bytes) and cannot be chunked automatically`,
        );
      }
    } else {
      // Store normally if small enough
      await redis.setEx(key, REDIS_TTL, dataString);
      console.log(
        `Redis: Stored wind data at key '${key}' with TTL ${REDIS_TTL}s`,
      );
    }
  } catch (error) {
    console.error("Redis: Error storing wind data:", error);
    throw error;
  }
}

/**
 * Get wind data from Redis (with automatic chunk reassembly)
 * @param {string} key - Redis key (default: 'wind:data')
 * @returns {Object|Array|null} - Parsed wind data or null if not found
 */
async function getWindData(key = "wind:data") {
  try {
    const redis = await getClient();

    // Check if data is chunked
    const chunkCount = await redis.get(`${key}:chunks`);

    if (chunkCount) {
      // Check if there's metadata (for objects with chunked points array)
      const metaData = await redis.get(`${key}:meta`);

      // Retrieve all chunks
      const numChunks = parseInt(chunkCount, 10);
      console.log(`Redis: Retrieving ${numChunks} chunks from key '${key}'...`);

      const chunks = [];
      for (let i = 0; i < numChunks; i++) {
        const chunkData = await redis.get(`${key}:chunk:${i}`);
        if (chunkData) {
          chunks.push(JSON.parse(chunkData));
        }
      }

      // Merge chunks
      const points = chunks.flat();

      if (metaData) {
        // Reconstruct object with points
        const metadata = JSON.parse(metaData);
        metadata.points = points;
        console.log(
          `Redis: Retrieved and merged ${numChunks} chunks (${points.length} points) from key '${key}'`,
        );
        return metadata;
      } else {
        // Return merged array
        console.log(
          `Redis: Retrieved and merged ${numChunks} chunks (${points.length} items) from key '${key}'`,
        );
        return points;
      }
    } else {
      // Retrieve normal (non-chunked) data
      const data = await redis.get(key);
      if (!data) {
        console.log(`Redis: No data found at key '${key}'`);
        return null;
      }
      console.log(`Redis: Retrieved wind data from key '${key}'`);
      return JSON.parse(data);
    }
  } catch (error) {
    console.error("Redis: Error retrieving wind data:", error);
    throw error;
  }
}

/**
 * Store binary data (PNG image) in Redis
 * @param {Buffer} buffer - Binary data
 * @param {string} key - Redis key
 */
async function setBinaryData(buffer, key) {
  try {
    const redis = await getClient();
    // Store as base64 to avoid encoding issues
    const base64Data = buffer.toString("base64");
    await redis.setEx(key, REDIS_TTL, base64Data);
    console.log(
      `Redis: Stored binary data at key '${key}' (${buffer.length} bytes) with TTL ${REDIS_TTL}s`,
    );
  } catch (error) {
    console.error("Redis: Error storing binary data:", error);
    throw error;
  }
}

/**
 * Get binary data from Redis
 * @param {string} key - Redis key
 * @returns {Buffer|null} - Buffer or null if not found
 */
async function getBinaryData(key) {
  try {
    const redis = await getClient();

    // Get base64 encoded data
    const base64Data = await redis.get(key);

    if (!base64Data) {
      console.log(`Redis: No binary data found at key '${key}'`);
      return null;
    }

    // Decode from base64 to Buffer
    const buffer = Buffer.from(base64Data, "base64");
    console.log(
      `Redis: Retrieved binary data from key '${key}' (${buffer.length} bytes)`,
    );
    return buffer;
  } catch (error) {
    console.error("Redis: Error retrieving binary data:", error);
    throw error;
  }
}

/**
 * Check if Redis is connected
 */
function isRedisConnected() {
  return isConnected;
}

/**
 * Close Redis connection
 */
async function closeRedis() {
  if (client) {
    await client.quit();
    client = null;
    isConnected = false;
    console.log("Redis: Connection closed gracefully");
  }
}

/**
 * Store wind data with index for historical tracking
 * Maintains the last N versions of wind data
 * @param {Object|Array} data - Wind data to store
 * @param {string} baseKey - Base Redis key (e.g., 'wind:points')
 * @param {number} maxHistory - Maximum number of historical entries to keep (default: 10)
 */
async function setWindDataWithIndex(data, baseKey, maxHistory = 10) {
  try {
    const redis = await getClient();

    // Get current index, or start at 0
    const currentIndexStr = await redis.get(`${baseKey}:current_index`);
    let currentIndex = currentIndexStr ? parseInt(currentIndexStr, 10) : 0;

    // Get existing indices list
    const indicesStr = await redis.get(`${baseKey}:indices`);
    let indices = indicesStr ? JSON.parse(indicesStr) : [];

    // Store the new data with current index
    const indexedKey = `${baseKey}:${currentIndex}`;
    await setWindData(data, indexedKey);

    // Add timestamp info to indices list
    const indexEntry = {
      index: currentIndex,
      timestamp: new Date().toISOString(),
      dataPoints: Array.isArray(data)
        ? data.length
        : data.points
          ? data.points.length
          : 0,
      // Include historical data metadata if available
      runName: data.runName || null, // GFS run identifier (e.g., "20260121_00Z")
      dataTime: data.dataTime || null,
      hoursBack: data.hoursBack || null,
      forecastOffset: data.forecastOffset || null,
      runAge: data.runAge || null,
    };

    // Add new index to the list
    indices.push(indexEntry);

    // Keep only the last maxHistory entries
    if (indices.length > maxHistory) {
      const oldIndices = indices.slice(0, indices.length - maxHistory);

      // Delete old data
      for (const oldIndex of oldIndices) {
        const oldKey = `${baseKey}:${oldIndex.index}`;

        // Delete chunked data if exists
        const chunkCount = await redis.get(`${oldKey}:chunks`);
        if (chunkCount) {
          const numChunks = parseInt(chunkCount, 10);
          for (let i = 0; i < numChunks; i++) {
            await redis.del(`${oldKey}:chunk:${i}`);
          }
          await redis.del(`${oldKey}:chunks`);
          await redis.del(`${oldKey}:meta`);
        } else {
          await redis.del(oldKey);
        }

        console.log(`Redis: Deleted old data at index ${oldIndex.index}`);
      }

      // Keep only recent indices
      indices = indices.slice(indices.length - maxHistory);
    }

    // Store updated indices list
    await redis.setEx(`${baseKey}:indices`, REDIS_TTL, JSON.stringify(indices));

    // Update current index for next time
    const nextIndex = currentIndex + 1;
    await redis.setEx(
      `${baseKey}:current_index`,
      REDIS_TTL,
      nextIndex.toString(),
    );

    // Also store as latest (backward compatibility)
    await setWindData(data, baseKey);

    console.log(
      `Redis: Stored data at index ${currentIndex}, total history: ${indices.length}`,
    );
    return currentIndex;
  } catch (error) {
    console.error("Redis: Error storing indexed wind data:", error);
    throw error;
  }
}

/**
 * Get wind data by index
 * @param {string} baseKey - Base Redis key
 * @param {number} index - Index to retrieve
 * @returns {Object|Array|null}
 */
async function getWindDataByIndex(baseKey, index) {
  try {
    const indexedKey = `${baseKey}:${index}`;
    return await getWindData(indexedKey);
  } catch (error) {
    console.error(
      `Redis: Error retrieving wind data at index ${index}:`,
      error,
    );
    throw error;
  }
}

/**
 * Get binary data by index
 * @param {string} baseKey - Base Redis key
 * @param {number} index - Index to retrieve
 * @returns {Buffer|null}
 */
async function getBinaryDataByIndex(baseKey, index) {
  try {
    const indexedKey = `${baseKey}:${index}`;
    return await getBinaryData(indexedKey);
  } catch (error) {
    console.error(
      `Redis: Error retrieving binary data at index ${index}:`,
      error,
    );
    throw error;
  }
}

/**
 * Get list of available indices with timestamps
 * @param {string} baseKey - Base Redis key
 * @returns {Array}
 */
async function getAvailableIndices(baseKey) {
  try {
    const redis = await getClient();
    const indicesStr = await redis.get(`${baseKey}:indices`);

    if (!indicesStr) {
      return [];
    }

    const indices = JSON.parse(indicesStr);

    // Sort by dataTime (oldest first, most recent last)
    indices.sort((a, b) => {
      const dateA = new Date(a.dataTime).getTime();
      const dateB = new Date(b.dataTime).getTime();
      return dateB - dateA; // Ascending order (oldest first)
    });

    return indices;
  } catch (error) {
    console.error("Redis: Error retrieving available indices:", error);
    throw error;
  }
}

/**
 * Get the latest index number
 * @param {string} baseKey - Base Redis key
 * @returns {number|null}
 */
async function getLatestIndex(baseKey) {
  try {
    const indices = await getAvailableIndices(baseKey);

    if (indices.length === 0) {
      return null;
    }

    // Return the last index in the array (most recent)
    return indices[indices.length - 1].index;
  } catch (error) {
    console.error("Redis: Error retrieving latest index:", error);
    throw error;
  }
}

/**
 * Store binary data with index for historical tracking
 * @param {Buffer} buffer - Binary data to store
 * @param {string} baseKey - Base Redis key
 * @param {number} index - Index to use (should match the wind data index)
 */
async function setBinaryDataWithIndex(buffer, baseKey, index) {
  try {
    const indexedKey = `${baseKey}:${index}`;
    await setBinaryData(buffer, indexedKey);

    // Also store as latest (backward compatibility)
    await setBinaryData(buffer, baseKey);

    console.log(`Redis: Stored binary data at index ${index}`);
    return index;
  } catch (error) {
    console.error("Redis: Error storing indexed binary data:", error);
    throw error;
  }
}

module.exports = {
  initRedis,
  getClient,
  setWindData,
  getWindData,
  setBinaryData,
  getBinaryData,
  setWindDataWithIndex,
  setBinaryDataWithIndex,
  getWindDataByIndex,
  getBinaryDataByIndex,
  getAvailableIndices,
  getLatestIndex,
  isRedisConnected,
  closeRedis,
};

File: src/services/redis_client.rs
use anyhow::{Context, Result};
use base64::{engine::general_purpose, Engine as _};
use redis::{aio::ConnectionManager, AsyncCommands, Client};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tracing::info;

const REDIS_TTL: u64 = 60 * 60; // 1 hour in seconds
const MAX_SIZE: usize = 8 * 1024 * 1024; // 8 MB

#[derive(Clone)]
pub struct RedisClient {
    conn: Arc<ConnectionManager>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndexEntry {
    pub index: u32,
    pub timestamp: String,
    #[serde(rename = "dataPoints")]
    pub data_points: usize,
    #[serde(rename = "runName")]
    pub run_name: Option<String>,
    #[serde(rename = "dataTime")]
    pub data_time: Option<String>,
    #[serde(rename = "hoursBack")]
    pub hours_back: Option<f64>,
    #[serde(rename = "forecastOffset")]
    pub forecast_offset: Option<i32>,
    #[serde(rename = "runAge")]
    pub run_age: Option<String>,
}

impl RedisClient {
    pub async fn new(redis_url: &str) -> Result<Self> {
        info!("Connecting to Redis at {}", redis_url);

        let client = Client::open(redis_url)
            .context("Failed to create Redis client")?;

        let conn = ConnectionManager::new(client)
            .await
            .context("Failed to connect to Redis")?;

        info!("Redis: Connected and ready");

        Ok(Self {
            conn: Arc::new(conn),
        })
    }

    /// Store wind data in Redis with automatic chunking for large datasets
    pub async fn set_wind_data(&self, data: &serde_json::Value, key: &str) -> Result<()> {
        let data_string = serde_json::to_string(data)?;
        let data_size = data_string.as_bytes().len();

        if data_size > MAX_SIZE {
            // Check if data is an array or object with large 'points' property
            if let Some(arr) = data.as_array() {
                self.store_chunked_array(arr, key).await?;
            } else if let Some(obj) = data.as_object() {
                if let Some(points) = obj.get("points").and_then(|p| p.as_array()) {
                    self.store_chunked_object(obj, points, key).await?;
                } else {
                    anyhow::bail!(
                        "Data too large ({} bytes) and cannot be chunked automatically",
                        data_size
                    );
                }
            } else {
                anyhow::bail!(
                    "Data too large ({} bytes) and cannot be chunked automatically",
                    data_size
                );
            }
        } else {
            // Store normally if small enough
            let mut conn = self.conn.as_ref().clone();
            conn.set_ex::<_, _, ()>(key, data_string, REDIS_TTL).await?;
            info!("Redis: Stored wind data at key '{}' with TTL {}s", key, REDIS_TTL);
        }

        Ok(())
    }

    async fn store_chunked_array(&self, arr: &[serde_json::Value], key: &str) -> Result<()> {
        let data_string = serde_json::to_string(arr)?;
        let data_size = data_string.as_bytes().len();

        let chunk_size = (arr.len() as f64 / (data_size as f64 / MAX_SIZE as f64).ceil()).ceil() as usize;
        let chunks: Vec<&[serde_json::Value]> = arr.chunks(chunk_size).collect();

        info!(
            "Redis: Array too large ({} bytes), splitting into {} chunks...",
            data_size,
            chunks.len()
        );

        let mut conn = self.conn.as_ref().clone();

        // Store chunk count
        conn.set_ex::<_, _, ()>(
            format!("{}:chunks", key),
            chunks.len().to_string(),
            REDIS_TTL,
        )
        .await?;

        // Store each chunk
        for (i, chunk) in chunks.iter().enumerate() {
            let chunk_key = format!("{}:chunk:{}", key, i);
            let chunk_string = serde_json::to_string(chunk)?;
            conn.set_ex::<_, _, ()>(chunk_key, chunk_string, REDIS_TTL).await?;
        }

        info!(
            "Redis: Stored {} items in {} chunks at key '{}' with TTL {}s",
            arr.len(),
            chunks.len(),
            key,
            REDIS_TTL
        );

        Ok(())
    }

    async fn store_chunked_object(
        &self,
        obj: &serde_json::Map<String, serde_json::Value>,
        points: &[serde_json::Value],
        key: &str,
    ) -> Result<()> {
        let data_string = serde_json::to_string(obj)?;
        let data_size = data_string.as_bytes().len();

        let chunk_size = (points.len() as f64 / (data_size as f64 / MAX_SIZE as f64).ceil()).ceil() as usize;
        let chunks: Vec<&[serde_json::Value]> = points.chunks(chunk_size).collect();

        info!(
            "Redis: Object with large points array ({} bytes), splitting points into {} chunks...",
            data_size,
            chunks.len()
        );

        // Create metadata without points
        let mut meta = obj.clone();
        meta.remove("points");

        let mut conn = self.conn.as_ref().clone();

        // Store metadata
        conn.set_ex::<_, _, ()>(
            format!("{}:meta", key),
            serde_json::to_string(&meta)?,
            REDIS_TTL,
        )
        .await?;

        // Store chunk count
        conn.set_ex::<_, _, ()>(
            format!("{}:chunks", key),
            chunks.len().to_string(),
            REDIS_TTL,
        )
        .await?;

        // Store each chunk
        for (i, chunk) in chunks.iter().enumerate() {
            let chunk_key = format!("{}:chunk:{}", key, i);
            let chunk_string = serde_json::to_string(chunk)?;
            conn.set_ex::<_, _, ()>(chunk_key, chunk_string, REDIS_TTL).await?;
        }

        info!(
            "Redis: Stored {} points in {} chunks at key '{}' with TTL {}s",
            points.len(),
            chunks.len(),
            key,
            REDIS_TTL
        );

        Ok(())
    }

    /// Get wind data from Redis with automatic chunk reassembly
    pub async fn get_wind_data(&self, key: &str) -> Result<Option<serde_json::Value>> {
        let mut conn = self.conn.as_ref().clone();

        // Check if data is chunked
        let chunk_count: Option<String> = conn.get(format!("{}:chunks", key)).await?;

        if let Some(chunk_count_str) = chunk_count {
            let num_chunks: usize = chunk_count_str.parse()?;

            info!("Redis: Retrieving {} chunks from key '{}'...", num_chunks, key);

            // Check if there's metadata
            let meta_data: Option<String> = conn.get(format!("{}:meta", key)).await?;

            // Retrieve all chunks
            let mut points = Vec::new();
            for i in 0..num_chunks {
                let chunk_key = format!("{}:chunk:{}", key, i);
                let chunk_data: Option<String> = conn.get(&chunk_key).await?;

                if let Some(chunk_str) = chunk_data {
                    let chunk: Vec<serde_json::Value> = serde_json::from_str(&chunk_str)?;
                    points.extend(chunk);
                }
            }

            if let Some(meta_str) = meta_data {
                // Reconstruct object with points
                let mut metadata: serde_json::Map<String, serde_json::Value> =
                    serde_json::from_str(&meta_str)?;
                metadata.insert("points".to_string(), serde_json::json!(points));

                info!(
                    "Redis: Retrieved and merged {} chunks ({} points) from key '{}'",
                    num_chunks,
                    points.len(),
                    key
                );

                Ok(Some(serde_json::Value::Object(metadata)))
            } else {
                // Return merged array
                info!(
                    "Redis: Retrieved and merged {} chunks ({} items) from key '{}'",
                    num_chunks,
                    points.len(),
                    key
                );

                Ok(Some(serde_json::json!(points)))
            }
        } else {
            // Retrieve normal (non-chunked) data
            let data: Option<String> = conn.get(key).await?;

            if let Some(data_str) = data {
                info!("Redis: Retrieved wind data from key '{}'", key);
                Ok(Some(serde_json::from_str(&data_str)?))
            } else {
                info!("Redis: No data found at key '{}'", key);
                Ok(None)
            }
        }
    }

    /// Store binary data (PNG image) in Redis with base64 encoding
    pub async fn set_binary_data(&self, buffer: &[u8], key: &str) -> Result<()> {
        let base64_data = general_purpose::STANDARD.encode(buffer);

        let mut conn = self.conn.as_ref().clone();
        conn.set_ex::<_, _, ()>(key, base64_data, REDIS_TTL).await?;

        info!(
            "Redis: Stored binary data at key '{}' ({} bytes) with TTL {}s",
            key,
            buffer.len(),
            REDIS_TTL
        );

        Ok(())
    }

    /// Get binary data from Redis with base64 decoding
    pub async fn get_binary_data(&self, key: &str) -> Result<Option<Vec<u8>>> {
        let mut conn = self.conn.as_ref().clone();
        let base64_data: Option<String> = conn.get(key).await?;

        if let Some(data_str) = base64_data {
            let buffer = general_purpose::STANDARD.decode(&data_str)?;

            info!(
                "Redis: Retrieved binary data from key '{}' ({} bytes)",
                key,
                buffer.len()
            );

            Ok(Some(buffer))
        } else {
            info!("Redis: No binary data found at key '{}'", key);
            Ok(None)
        }
    }

    /// Store wind data with index for historical tracking
    pub async fn set_wind_data_with_index(
        &self,
        data: &serde_json::Value,
        base_key: &str,
        max_history: usize,
    ) -> Result<u32> {
        let mut conn = self.conn.as_ref().clone();

        // Get current index
        let current_index_str: Option<String> = conn.get(format!("{}:current_index", base_key)).await?;
        let mut current_index: u32 = current_index_str
            .and_then(|s| s.parse().ok())
            .unwrap_or(0);

        // Get existing indices list
        let indices_str: Option<String> = conn.get(format!("{}:indices", base_key)).await?;
        let mut indices: Vec<IndexEntry> = indices_str
            .and_then(|s| serde_json::from_str(&s).ok())
            .unwrap_or_default();

        // Extract metadata from data
        let data_time = data.get("dataTime").and_then(|v| v.as_str()).map(String::from);
        let run_name = data.get("runName").and_then(|v| v.as_str()).map(String::from);
        let hours_back = data.get("hoursBack").and_then(|v| v.as_f64());
        let forecast_offset = data.get("forecastOffset").and_then(|v| v.as_i64()).map(|v| v as i32);
        let run_age = data.get("runAge").and_then(|v| v.as_str()).map(String::from);

        // Check for duplicate entry (within 2h tolerance)
        let tolerance_ms = 2 * 60 * 60 * 1000; // 2 hours
        let existing_entry_index = if let Some(dt) = &data_time {
            if let Ok(data_time_ms) = chrono::DateTime::parse_from_rfc3339(dt) {
                indices.iter().position(|idx| {
                    if let Some(existing_dt) = &idx.data_time {
                        if let Ok(existing_time_ms) = chrono::DateTime::parse_from_rfc3339(existing_dt) {
                            let diff = (data_time_ms.timestamp_millis() - existing_time_ms.timestamp_millis()).abs();
                            return diff < tolerance_ms;
                        }
                    }
                    false
                })
            } else {
                None
            }
        } else {
            None
        };

        // Update or create entry
        if let Some(idx_pos) = existing_entry_index {
            let existing_entry = &indices[idx_pos];
            info!(
                "Redis: Updating existing entry at index {} for {:?}",
                existing_entry.index, data_time
            );
            current_index = existing_entry.index;

            indices[idx_pos] = IndexEntry {
                index: current_index,
                timestamp: chrono::Utc::now().to_rfc3339(),
                data_points: Self::count_data_points(data),
                run_name: run_name.clone(),
                data_time: data_time.clone(),
                hours_back,
                forecast_offset,
                run_age: run_age.clone(),
            };
        } else {
            // Create new entry
            let index_entry = IndexEntry {
                index: current_index,
                timestamp: chrono::Utc::now().to_rfc3339(),
                data_points: Self::count_data_points(data),
                run_name: run_name.clone(),
                data_time: data_time.clone(),
                hours_back,
                forecast_offset,
                run_age: run_age.clone(),
            };

            indices.push(index_entry);
        }

        // Store the data with the index
        let indexed_key = format!("{}:{}", base_key, current_index);
        self.set_wind_data(data, &indexed_key).await?;

        // Keep only the last maxHistory entries
        if indices.len() > max_history {
            let old_indices: Vec<_> = indices.drain(..indices.len() - max_history).collect();

            // Delete old data
            for old_index in old_indices {
                let old_key = format!("{}:{}", base_key, old_index.index);
                self.delete_data(&old_key).await?;
                info!("Redis: Deleted old data at index {}", old_index.index);
            }
        }

        // Store updated indices list
        conn.set_ex::<_, _, ()>(
            format!("{}:indices", base_key),
            serde_json::to_string(&indices)?,
            REDIS_TTL,
        )
        .await?;

        // Update current index for next time (only if we created a new entry)
        if existing_entry_index.is_none() {
            let next_index = current_index + 1;
            conn.set_ex::<_, _, ()>(
                format!("{}:current_index", base_key),
                next_index.to_string(),
                REDIS_TTL,
            )
            .await?;
        }

        // Also store as latest (backward compatibility)
        self.set_wind_data(data, base_key).await?;

        info!(
            "Redis: Stored data at index {}, total history: {}",
            current_index,
            indices.len()
        );

        Ok(current_index)
    }

    /// Store binary data with index
    pub async fn set_binary_data_with_index(
        &self,
        buffer: &[u8],
        base_key: &str,
        index: u32,
    ) -> Result<()> {
        let indexed_key = format!("{}:{}", base_key, index);
        self.set_binary_data(buffer, &indexed_key).await?;

        // Also store as latest (backward compatibility)
        self.set_binary_data(buffer, base_key).await?;

        info!("Redis: Stored binary data at index {}", index);
        Ok(())
    }

    /// Get wind data by index
    pub async fn get_wind_data_by_index(
        &self,
        base_key: &str,
        index: u32,
    ) -> Result<Option<serde_json::Value>> {
        let indexed_key = format!("{}:{}", base_key, index);
        self.get_wind_data(&indexed_key).await
    }

    /// Get binary data by index
    pub async fn get_binary_data_by_index(
        &self,
        base_key: &str,
        index: u32,
    ) -> Result<Option<Vec<u8>>> {
        let indexed_key = format!("{}:{}", base_key, index);
        self.get_binary_data(&indexed_key).await
    }

    /// Get list of available indices with timestamps
    pub async fn get_available_indices(&self, base_key: &str) -> Result<Vec<IndexEntry>> {
        let mut conn = self.conn.as_ref().clone();
        let indices_str: Option<String> = conn.get(format!("{}:indices", base_key)).await?;

        if let Some(indices_json) = indices_str {
            let mut indices: Vec<IndexEntry> = serde_json::from_str(&indices_json)?;

            // Sort by dataTime (newest first, oldest last)
            indices.sort_by(|a, b| {
                let date_a = a.data_time.as_ref()
                    .and_then(|dt| chrono::DateTime::parse_from_rfc3339(dt).ok())
                    .map(|dt| dt.timestamp_millis())
                    .unwrap_or(0);
                let date_b = b.data_time.as_ref()
                    .and_then(|dt| chrono::DateTime::parse_from_rfc3339(dt).ok())
                    .map(|dt| dt.timestamp_millis())
                    .unwrap_or(0);
                date_b.cmp(&date_a) // Descending order (newest first)
            });

            Ok(indices)
        } else {
            Ok(Vec::new())
        }
    }

    /// Helper to delete data (including chunks)
    async fn delete_data(&self, key: &str) -> Result<()> {
        let mut conn = self.conn.as_ref().clone();

        // Check if chunked
        let chunk_count: Option<String> = conn.get(format!("{}:chunks", key)).await?;

        if let Some(chunk_count_str) = chunk_count {
            let num_chunks: usize = chunk_count_str.parse()?;

            for i in 0..num_chunks {
                let _: () = conn.del(format!("{}:chunk:{}", key, i)).await?;
            }

            let _: () = conn.del(format!("{}:chunks", key)).await?;
            let _: () = conn.del(format!("{}:meta", key)).await?;
        } else {
            let _: () = conn.del(key).await?;
        }

        Ok(())
    }

    /// Helper to count data points
    fn count_data_points(data: &serde_json::Value) -> usize {
        if let Some(arr) = data.as_array() {
            arr.len()
        } else if let Some(obj) = data.as_object() {
            obj.get("points")
                .and_then(|p| p.as_array())
                .map(|a| a.len())
                .unwrap_or(0)
        } else {
            0
        }
    }
}
